### PR TITLE
Switch CoreCLR finalizer loop to C#

### PR DESCRIPTION
**Pull request by jkotas**
Context: https://github.com/dotnet/runtime/pull/103425#issuecomment-2168534480

### CODE REVIEW

**Comment by jkotas** [comment]
@janvorli Could you please run your ad-hoc micro-benchmark (https://github.com/dotnet/runtime/pull/103425#issuecomment-2169002812) on this change?

**Comment by janvorli** [comment]
@jkotas the performance is about the same as before this change (very slightly worse, 2-5%).

File: src/coreclr/System.Private.CoreLib/src/System/Threading/Thread.CoreCLR.cs [hunk 1]
```diff
@@ -361,14 +361,39 @@ internal static int OptimalMaxSpinWaitsPerSpinIteration
...
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private void ResetFinalizerThreadSlow()
+        {
+            Debug.Assert(this == CurrentThread);
+            Debug.Assert(_mayNeedResetForThreadPool);
+
+            _mayNeedResetForThreadPool = false;
+
+            const string FinalizerThreadName = ".NET Finalizer";
```
Comment (janvorli) [inline src/coreclr/System.Private.CoreLib/src/System/Threading/Thread.CoreCLR.cs:382]: I am not sure I understand why we are doing this rename. Aren't we always calling the RunFinalizers on the finalizer thread that is already named as such?

**Review by janvorli — APPROVED** [review]
LGTM maybe modulo the comment, thank you!

File: src/coreclr/System.Private.CoreLib/src/System/Threading/Thread.CoreCLR.cs (hunk 1 above)
Comment (jkotas) [reply src/coreclr/System.Private.CoreLib/src/System/Threading/Thread.CoreCLR.cs:382]: User code can change the thread name. This is going to set it back if that happens.

Historically, threadpool threads and finalizer thread had code that tried to reset a few properties like name if they got modified by the user code. I never liked that behavior. There are many other ways that the user code can damage thread and finalizers threads that we are not accounting for.

The unmanaged version that I am deleting in this PR had a bug: https://github.com/dotnet/runtime/pull/103501/files#diff-f5835c4b5fd134e52b4127bb4ffb7e5ad439673a429dc7ea46d53e7a5bca0529L7735 . It reset the finalizer thread name to NULL, not the name that we use for finalizer threads.

**Comment by stephentoub** [comment]
For my edification, what does this buy us? just the general "more C# rather than C is better" charge?

**Comment by jkotas** [comment]
- It fixes a bug in new exception handling.
- It reduces finalizer loop overhead on Unix and Windows Arm64 (~10+% for no-op finalizer)

Before this change, Windows x86/x64 had hand-written assembly code to minimize finalizer loop overhead. The exception handling bug was related to this hand-written assembly code. Switching to managed implementation allowed us to fix the exception handling bug, get rid of the x86/x64 had hand-written assembly code, and improve perf outside Windows x86/x64.

An alternative fix would be to keep writing more platform-specific code to fix the exception handling bug, and potentially keep writing even more platform-specific code to get other platforms to perf parity with Windows x86/x64.

**Comment by jkotas** [comment]
https://github.com/dotnet/runtime/issues/103385 is the EH issue that this contributes to.

**Comment by stephentoub** [comment]
Excellent, thanks

**Pull request merged**